src/System.Slices/src/System/Span.cs

```diff
@@ -0,0 +1,293 @@
+        /// Thrown if the 'array' parameter is null.
+        /// </exception>
+        public Span(T[] array)
+        {
+            Contract.Requires(array != null);
+            m_object = array;
+            m_offset = new UIntPtr((uint)SpanHelpers<T>.OffsetToArrayData);
+            Length = array.Length;
+        }
+
+        /// <summary>
+        /// Creates a new span over the portion of the target array beginning
+        /// at 'start' index.
+        /// </summary>
+        /// <param name="array">The target array.</param>
+        /// <param name="start">The index at which to begin the span.</param>
+        /// <exception cref="System.ArgumentException">
+        /// Thrown if the 'array' parameter is null.
+        /// </exception>
+        /// <exception cref="System.ArgumentOutOfRangeException">
+        /// Thrown when the specified start index is not in range (&lt;0 or &gt;&eq;length).
+        /// </exception>
+        public Span(T[] array, int start)
+        {
+            Contract.Requires(array != null);
+            Contract.RequiresInInclusiveRange(start, array.Length);
+            if (start < array.Length) {
+                m_object = array;
+                m_offset = new UIntPtr(
+                    (uint)(SpanHelpers<T>.OffsetToArrayData + (start * PtrUtils.SizeOf<T>())));
+                Length = array.Length - start;
+            }
+            else {
+                m_object = null;
+                m_offset = UIntPtr.Zero;
+                Length = 0;
+            }
+        }
+
+        /// <summary>
+        /// Creates a new span over the portion of the target array beginning
+        /// at 'start' index and ending at 'end' index (exclusive).
+        /// </summary>
+        /// <param name="array">The target array.</param>
+        /// <param name="start">The index at which to begin the span.</param>
+        /// <param name="end">The index at which to end the span (exclusive).</param>
+        /// <exception cref="System.ArgumentException">
+        /// Thrown if the 'array' parameter is null.
+        /// </exception>
+        /// <exception cref="System.ArgumentOutOfRangeException">
+        /// Thrown when the specified start or end index is not in range (&lt;0 or &gt;&eq;length).
+        /// </exception>
+        public Span(T[] array, int start, int end)
+        {
+            Contract.Requires(array != null);
+            Contract.RequiresInInclusiveRange(start, array.Length);
+            if (start < array.Length) {
+                m_object = array;
+                m_offset = new UIntPtr(
+                    (uint)(SpanHelpers<T>.OffsetToArrayData + (start * PtrUtils.SizeOf<T>())));
+                Length = end - start;
+            }
+            else {
+                m_object = null;
+                m_offset = UIntPtr.Zero;
+                Length = 0;
+            }
+        }
+
+        /// <summary>
+        /// Creates a new span over the target unmanaged buffer.  Clearly this
```

Comment:
BTW, I did not write the original comment (@joeduffy did). But I have to say I like Joe's better. It explains what constitutes a valid buffer (right size for Ts). Your version just says that it needs to be valid.